activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -12,13 +12,15 @@ def records_by_owner
           return @records_by_owner if defined?(@records_by_owner)
 
           @records_by_owner = owners.each_with_object({}) do |owner, result|
+            through_records = through_records_by_owner[owner] || []
+
             if loaded?(owner)
-              result[owner] = target_for(owner)
+              result[owner] = through_records.flat_map do |record|
+                source_records_by_owner[record] || []
+              end
```

Comment:
When the intermediate record was preloaded, result[owner] was set with a new `target_for(owner)` instead of using the already loaded records.

This resulted in a different objects for:
- the values of `records_by_owner` of the through objects
- the keys of `records_by_owner` for the source records
which were expected to be the same.